### PR TITLE
cop: change handle_request interface to async

### PR DESCRIPTION
**Pull request by sticnarf**
###  What have you changed?

This PR simplifies `handle_unary_request_impl` using async/await. It also changes `handle_request` to an async function so that we may interrupt it in the future.

Due to limitations of async/await and future compat layer, this PR introduces two more allocations. Benchmark shows there's no significant influence.

sysbench 32 tables x 1M rows
workload: read_only without point get (all requests are small coprocessor requests)

Master:

| threads | qps | p95(ms) |
|--------|--------|-----------
| 32 | 23117.94 | 11.65 |
| 128 | 27036.21 | 45.79 |

This PR:

| threads | qps | p95(ms) |
|--------|--------|-----------
| 32 | 23181.34 | 11.65 |
| 128 | 26904.93 | 45.79 |

###  What is the type of the changes?

- Engineering (engineering change which doesn't change any feature or fix any issue)

###  How is the PR tested?

- Unit test
- Integration test

###  Does this PR affect documentation (docs) or should it be mentioned in the release notes?

No

###  Does this PR affect `tidb-ansible`?

No

### CODE REVIEW

File: src/coprocessor/endpoint.rs [hunk 1]
```diff
@@ -293,69 +294,72 @@ impl<E: Engine> Endpoint<E> {
...
+            result.map_err(Error::from)
+        }
+    }
+
     /// The real implementation of handling a unary request.
     ///
     /// It first retrieves a snapshot, then builds the `RequestHandler` over the snapshot and
     /// the given `handler_builder`. Finally, it calls the unary request interface of the
     /// `RequestHandler` to process the request and produce a result.
     // TODO: Convert to use async / await.
```
Comment (sticnarf) [inline src/coprocessor/endpoint.rs]: should be removed

**Comment by sticnarf** [comment]
@breeswish Ok. I'll also do it for the streaming part.

**Comment by sticnarf** [comment]
@breeswish I've changed the streaming part to async/await too. PTAL.

File: components/tikv_util/src/future.rs [hunk 2]
```diff
@@ -20,6 +20,23 @@ where
     (callback, future)
 }
 
+pub fn paired_future_callback_new<T>() -> (
```
[the remark this answers is not in the document]
Comment (sticnarf) [reply components/tikv_util/src/future.rs]: I change it to `paired_std_future_callback`

File: src/coprocessor/endpoint.rs [hunk 3]
```diff
@@ -281,81 +282,73 @@ impl<E: Engine> Endpoint<E> {
...
-            })
-            .and_then(|(_ctx, result)| result)
-            // map engine::Error -> coprocessor::Error
-            .map_err(Error::from)
+        // make engine not cross yield point
+        async move {
+            val?; // propagate error
+            let (_ctx, result) = future
+                .map_err(|cancel| KvError::from(KvErrorInner::Other(box_err!(cancel))))
+                .await?;
```
Comment (breeswish) [inline src/coprocessor/endpoint.rs]: Why after `?` there is still a `result`?

Comment (sticnarf) [reply src/coprocessor/endpoint.rs]: The `?` propagates the oneshot channel error. And the `result` is from what the callback returns:

`pub type Callback<T> = Box<dyn FnOnce((CbContext, Result<T>)) + Send>;`

**Pull request merged**